mqtt/mqtt-broker/tests/smoke.rs

```diff
@@ -113,3 +115,153 @@ async fn basic_pub_sub() {
+        .expect("couldn't publish");
+
+    client
+        .publish(Publication {
+            topic_name: topic_b.into(),
+            qos: QoS::AtLeastOnce,
+            retain: true,
+            payload: "r qos 1".into(),
+        })
+        .await
+        .expect("couldn't publish");
+
+    client
+        .publish(Publication {
+            topic_name: topic_c.into(),
+            qos: QoS::ExactlyOnce,
+            retain: true,
+            payload: "r qos 2".into(),
+        })
+        .await
+        .expect("couldn't publish");
+
+    client
+        .subscribe(SubscribeTo {
+            topic_filter: "topic/+".into(),
+            qos: QoS::ExactlyOnce,
+        })
+        .await
+        .expect("couldn't subscribe to a topic");
+
+    client.next().await; //skip connect event
+
+    assert_matches!(
+        client.next().await,
+        Some(Event::SubscriptionUpdates(_))
+    );
+
+    let mut shutdown_handle = client.shutdown_handle();
+
+    // read and map 3 expected events from the stream
+    let mut events = client
+        .take(3)
+        .filter_map(|e| async {
+            match e {
+                Event::Publication(publication) => Some((publication.payload, publication.retain)),
+                _ => None,
+            }
+        })
+        .collect::<Vec<_>>()
+        .await;
+
+    // sort by payload for ease of comparison.
+    events.sort_by_key(|e| e.0.clone());
+
+    assert_eq!(3, events.len());
+    assert_eq!(events[0], (Bytes::from("r qos 0"), true));
+    assert_eq!(events[1], (Bytes::from("r qos 1"), true));
+    assert_eq!(events[2], (Bytes::from("r qos 2"), true));
+
+    shutdown_handle.shutdown().await.expect("couldn't shutdown");
+    broker_shutdown.send(()).expect("couldn't shutdown broker");
+    broker_task
+        .await
+        .unwrap()
+        .expect("can't wait for the broker");
+}
+
+#[tokio::test]
+async fn will_message() {
+    let topic = "topic/A";
+
+    let broker = BrokerBuilder::default()
+        .authenticator(|_| Ok(Some(AuthId::Anonymous)))
+        .authorizer(|_| Ok(true))
+        .build();
+
+    let (broker_shutdown, broker_task, address) = common::start_server(broker);
+
+    let mut client_a = TestClientBuilder::new(address.clone())
+        .client_id("mqtt-smoke-tests-a")
+        .will(Publication {
+            topic_name: topic.into(),
+            qos: QoS::AtLeastOnce,
+            retain: false,
+            payload: "will_msg_a".into(),
+        })
+        .keep_alive(Duration::from_secs(1))
+        .build();
+
+    client_a.next().await; //skip connect event
+
+    let mut client_b = TestClientBuilder::new(address.clone())
+        .client_id("mqtt-smoke-tests-b")
+        .build();
+
+    client_b
+        .subscribe(SubscribeTo {
+            topic_filter: topic.into(),
+            qos: QoS::ExactlyOnce,
+        })
+        .await
+        .expect("couldn't subscribe to a topic");
+
+    client_b.next().await; //skip connect event
```

Comment:
I'd be nice if we can avoid these events from tests

Reply:
just wonder: if we spit one stream into several streams by event type, will that work? like one stream for Connect event, one for SubscriptionUpdate, and one for Publish events? 
probably it can help use to await for connect when we need and just don't care about it when we dont need in tests. Also we can extract Event data to those streams...

Reply:
I like the idea, I'll explore it in the following PRs.